<a href="https://colab.research.google.com/github/alexgaskell10/NLP_Translation/blob/master/notebooks/baseline.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Coursework: Baseline Model

This notebook takes you step by step to the implementation of a simple baseline model to get you started on the coursework. You have a section for the English-German task and another for English-Chinese. They are made to be standalone so feel free to check only one of the sections. However, as the tasks require slighlty different approaches, going through both sections could help you to get inspired for your chosen task, especially each task processes english in a slighlty different way.

Enjoy!

## A. English-German

### Importing Data

In [1]:
# Download and unzip the data
from os.path import exists
if not exists('ende_data.zip'):
    !wget -O ende_data.zip https://competitions.codalab.org/my/datasets/download/c748d2c0-d6be-4e36-9f12-ca0e88819c4d
    !unzip ende_data.zip

--2020-02-07 08:35:26--  https://competitions.codalab.org/my/datasets/download/c748d2c0-d6be-4e36-9f12-ca0e88819c4d
Resolving competitions.codalab.org (competitions.codalab.org)... 129.175.22.230
Connecting to competitions.codalab.org (competitions.codalab.org)|129.175.22.230|:443... connected.
HTTP request sent, awaiting response... 302 FOUND
Location: https://newcodalab.lri.fr/prod-private/dataset_data_file/None/104ea/en-de.zip?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Expires=86400&X-Amz-SignedHeaders=host&X-Amz-Signature=edd0aa1a83e65dd1db791aed965a0edde71751b15380268ce049767cc36065dc&X-Amz-Date=20200207T083527Z&X-Amz-Credential=AZIAIOSAODNN7EX123LE%2F20200207%2Fnewcodalab%2Fs3%2Faws4_request [following]
--2020-02-07 08:35:27--  https://newcodalab.lri.fr/prod-private/dataset_data_file/None/104ea/en-de.zip?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Expires=86400&X-Amz-SignedHeaders=host&X-Amz-Signature=edd0aa1a83e65dd1db791aed965a0edde71751b15380268ce049767cc36065dc&X-Amz-Date=20200207T083

In [2]:
# Check the files
import io

#English-German
print("---EN-DE---")
print()

with open("./train.ende.src", "r") as ende_src:
    print("Source: ",ende_src.readline())
with open("./train.ende.mt", "r") as ende_mt:
    print("Translation: ",ende_mt.readline())
with open("./train.ende.scores", "r") as ende_scores:
    print("Score: ",ende_scores.readline())


---EN-DE---

Source:  José Ortega y Gasset visited Husserl at Freiburg in 1934.

Translation:  1934 besuchte José Ortega y Gasset Husserl in Freiburg.

Score:  1.1016968715664406



### Computing Sentence Embeddings 

For this baseline model, we will simply use pre-trained GloVe embeddings via the Spacy module and compute the vector for each word and take the global mean for each sentence. We will do the same for both source and translation sentences. For chinese tokenization and embeddings we will have to find other tools.

This is a very simplistic approach so feel free to be more creative and play around with how the sentence embeddings are computed for example ;).

GloVe embeddings do not support the Chinese language so in the section of the English-Chinese task we will have to download pretrained Chinese embeddings from word2vec repositories.

In [3]:
#Downloading spacy models for english and german

!spacy download en_core_web_md
!spacy link en_core_web_md en300

!spacy download de_core_news_md
!spacy link de_core_news_md de300

     |████████████████████████████████| 95.4MB 73.4MB/s 
  Created wheel for en-core-web-md: filename=en_core_web_md-2.1.0-cp36-none-any.whl size=97126236 sha256=dc9ae2c79d0806387537e7d3c9401e31a25f8db3e43711bd6561ff570a06e9ad
  Stored in directory: /tmp/pip-ephem-wheel-cache-j34p9wv6/wheels/c1/2c/5f/fd7f3ec336bf97b0809c86264d2831c5dfb00fc2e239d1bb01
Successfully built en-core-web-md
✔ Download and installation successful
You can now load the model via spacy.load('en_core_web_md')
✔ Linking successful
/usr/local/lib/python3.6/dist-packages/en_core_web_md -->
/usr/local/lib/python3.6/dist-packages/spacy/data/en300
You can now load the model via spacy.load('en300')
     |████████████████████████████████| 220.8MB 154.8MB/s 
  Created wheel for de-core-news-md: filename=de_core_news_md-2.1.0-cp36-none-any.whl size=224546880 sha256=499bf1246e222e74abea49f03c3bc06a7017957d3d1618fe70cba1f3ab061b03
  Stored in directory: /tmp/pip-ephem-wheel-cache-im4yi74g/wheels/44/34/f1/31d4b0fa32008c09695cc

We can now write our functions that will return the average embeddings for a sentence.

#### Pre-processing with Spacy

In [4]:
import numpy as np
import spacy

from nltk import download
from nltk.corpus import stopwords

#downloading stopwords from the nltk package
download('stopwords') #stopwords dictionary, run once

stop_words_en = set(stopwords.words('english'))
stop_words_de = set(stopwords.words('german'))

def get_sentence_emb(line,nlp,lang):
    if lang == 'en':
        text = line.lower()
        l = [token.lemma_ for token in nlp.tokenizer(text)]
        l = ' '.join([word for word in l if word not in stop_words_en])

    elif lang == 'de':
        text = line.lower()
        l = [token.lemma_ for token in nlp.tokenizer(text)]
        l = ' '.join([word for word in l if word not in stop_words_de])

    sen = nlp(l)
    return sen.vector

def get_embeddings(f,nlp,lang):
    file = open(f) 
    lines = file.readlines() 
    sentences_vectors =[]

    for l in lines:
        vec = get_sentence_emb(l,nlp,lang)
        if vec is not None:
            vec = np.mean(vec)
            sentences_vectors.append(vec)
        else:
            print("didn't work :", l)
            sentences_vectors.append(0)

    return sentences_vectors

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


#### Getting Training and Validation Sets

We will now run the code fo the English-German translations and getting our training and validation sets ready for the regression task.


In [ ]:
import spacy

nlp_de = spacy.load('de300')
nlp_en = spacy.load('en300')

In [33]:
import torch

#EN-DE files
de_train_src = get_embeddings("./train.ende.src",nlp_en,'en')
de_train_mt = get_embeddings("./train.ende.mt",nlp_de,'de')

f_train_scores = open("./train.ende.scores",'r')
de_train_scores = f_train_scores.readlines()

de_val_src = get_embeddings("./dev.ende.src",nlp_en,'en')
de_val_mt = get_embeddings("./dev.ende.mt",nlp_de,'de')
f_val_scores = open("./dev.ende.scores",'r')
de_val_scores = f_val_scores.readlines()

AttributeError: ignored

In [ ]:
#EN-DE
print(f"Training mt: {len(de_train_mt)} Training src: {len(de_train_src)}")
print()
print(f"Validation mt: {len(de_val_mt)} Validation src: {len(de_val_src)}")

In [ ]:
#Put the features into a list
X_train = [np.array(de_train_src),np.array(de_train_mt)]
X_train_de = np.array(X_train).transpose()

X_val = [np.array(de_val_src),np.array(de_val_mt)]
X_val_de = np.array(X_val).transpose()

#Scores
train_scores = np.array(de_train_scores).astype(float)
y_train_de =train_scores

val_scores = np.array(de_val_scores).astype(float)
y_val_de = val_scores

In [9]:
# RUN IF WANT TO HAVE AVERAGE VECTOR(AND NOT GLOBAL MEAN), THIS GAVE WORSE PERFORMANCE THAN GLOBAL MEAN
'''

X_train= [np.array(train_src),np.array(train_mt)]
X_train = np.array(X_train)


X_test = [np.array(test_src),np.array(test_mt)]
X_test = np.array(X_test)


#Reshaping if using shape >3
nsamples, nx, ny = X_train.shape
X_train = X_train.reshape((nx,ny*nsamples))

nsamples, nx, ny = X_test.shape
X_test = X_test.reshape((nx,ny*nsamples))

print(X_train.shape)
print(X_test.shape)



#Scores
train_scores = np.array(train_scores).astype(float)
y_train =train_scores

test_scores = np.array(test_scores).astype(float)
y_test =test_scores
'''

'\n\nX_train= [np.array(train_src),np.array(train_mt)]\nX_train = np.array(X_train)\n\n\nX_test = [np.array(test_src),np.array(test_mt)]\nX_test = np.array(X_test)\n\n\n#Reshaping if using shape >3\nnsamples, nx, ny = X_train.shape\nX_train = X_train.reshape((nx,ny*nsamples))\n\nnsamples, nx, ny = X_test.shape\nX_test = X_test.reshape((nx,ny*nsamples))\n\nprint(X_train.shape)\nprint(X_test.shape)\n\n\n\n#Scores\ntrain_scores = np.array(train_scores).astype(float)\ny_train =train_scores\n\ntest_scores = np.array(test_scores).astype(float)\ny_test =test_scores\n'

### Training the Regressor


At this point,  will try SVM and Random Tree Forests and choose the model with the highest Pearson correlation.

First we will define our RMSE function.

In [ ]:
import numpy as np

def rmse(predictions, targets):
    return np.sqrt(((predictions - targets) ** 2).mean())

#### SVM

SVM have many parameters such as the kernel and the regularizating constant C. Here we will use C = 1 and compare kernels. 

In [11]:
from sklearn.svm import SVR
from scipy.stats.stats import pearsonr

for k in ['linear','poly','rbf','sigmoid']:
    clf_t = SVR(kernel=k)
    clf_t.fit(X_train_de, y_train_de)
    print(k)
    predictions = clf_t.predict(X_val_de)
    pearson = pearsonr(y_val_de, predictions)
    print(f'RMSE: {rmse(predictions,y_val_de)} Pearson {pearson[0]}')
    print()

linear
RMSE: 0.8813940062646772 Pearson 0.05474933605816939

poly
RMSE: 0.881237789938125 Pearson 0.06200263088261973

rbf
RMSE: 0.881693995276586 Pearson 0.023417950678245963

sigmoid
RMSE: 349.7715466025825 Pearson -0.037125683901182906



Here the best kernel seems to be the polynomial one as it gives us the highest pearson correlation at 0.062.

#### Random Tree Forest

Another powerful regressor is the Random Tree Forest. Here we have to choose the number of trees we want to compute and we will pick n_estimators = 1000. The higher the number the longer it will compute. To fine tune that number you could compute the error per number of trees and select the number for which there is no more significant improvement( the "elbow" of the graph).

In [12]:
# Import the model we are using

from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(n_estimators = 1000, random_state = 666)

rf.fit(X_train_de, y_train_de);

predictions = rf.predict(X_val_de)

pearson = pearsonr(y_val_de, predictions)
print('RMSE:', rmse(predictions,y_val_de))
print(f"Pearson {pearson[0]}")

RMSE: 0.9401130737723049
Pearson -0.03680114854324954


In this case, it seems like the SVM with a linear kernel performed the best on our validation set so we will save that model for the test set predictions.

### Writing Results

Here is our function to write the scores into a txt file. We can follow the <Method> <ID> <SCORE> template but having only the scores will work too.

In [ ]:
import os

def writeScores(method_name, scores):
    fn = "predictions.txt"
    print("")
    with open(fn, 'w') as output_file:
        for idx,x in enumerate(scores):
            #out =  metrics[idx]+":"+str("{0:.2f}".format(x))+"\n"
            #print(out)
            output_file.write(f"{x}\n")

In [ ]:
#EN-DE

de_test_src = get_embeddings("./test.ende.src",nlp_en,'en')
de_test_mt = get_embeddings("./test.ende.mt",nlp_de,'de')

X= [np.array(de_test_src),np.array(de_test_mt)]
X_test = np.array(X).transpose()

#Predict
clf_de = SVR(kernel='rbf')
clf_de.fit(X_train_de, y_train_de)

predictions_de = clf_de.predict(X_val_de)

In [15]:
from google.colab import files
from zipfile import ZipFile

writeScores("SVR",predictions_de)

with ZipFile("en-de_svr.zip","w") as newzip:
    newzip.write("predictions.txt")
 
files.download('en-de_svr.zip') 

### Results

Once submitted to codalab, the pearson correlation is 0.0052.

##B. English-Chinese


### Importing Data

In [16]:
from os.path import exists

if not exists('enzh_data.zip'):
    !wget -O enzh_data.zip https://competitions.codalab.org/my/datasets/download/03e23bd7-8084-4542-997b-6a1ca6dd8a5f
    !unzip enzh_data.zip

--2020-02-07 08:42:15--  https://competitions.codalab.org/my/datasets/download/03e23bd7-8084-4542-997b-6a1ca6dd8a5f
Resolving competitions.codalab.org (competitions.codalab.org)... 129.175.22.230
Connecting to competitions.codalab.org (competitions.codalab.org)|129.175.22.230|:443... connected.
HTTP request sent, awaiting response... 302 FOUND
Location: https://newcodalab.lri.fr/prod-private/dataset_data_file/None/630ec/en-zh.zip?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Expires=86400&X-Amz-SignedHeaders=host&X-Amz-Signature=3add5877798fb6ad81b05bbb1a80936898d690b048df22de322955c53f92ad00&X-Amz-Date=20200207T084215Z&X-Amz-Credential=AZIAIOSAODNN7EX123LE%2F20200207%2Fnewcodalab%2Fs3%2Faws4_request [following]
--2020-02-07 08:42:15--  https://newcodalab.lri.fr/prod-private/dataset_data_file/None/630ec/en-zh.zip?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Expires=86400&X-Amz-SignedHeaders=host&X-Amz-Signature=3add5877798fb6ad81b05bbb1a80936898d690b048df22de322955c53f92ad00&X-Amz-Date=20200207T084

In [17]:
#English-Chinese
#Checking Data
print("---EN-ZH---")
print()

with open("./train.enzh.src", "r") as enzh_src:
    print("Source: ",enzh_src.readline())
with open("./train.enzh.mt", "r") as enzh_mt:
    print("Translation: ",enzh_mt.readline())
with open("./train.enzh.scores", "r") as enzh_scores:
    print("Score: ",enzh_scores.readline())

---EN-ZH---

Source:  The last conquistador then rides on with his sword drawn.

Translation:  最后的征服者骑着他的剑继续前进.

Score:  -1.5284005772625449



### Computing Sentence Embeddings


For this task, we will compute the embeddings for words in a sentence in one language and compute the global mean for that sentence, and do the same for the other language. However, we will have to find and download pre-traind embeddings for Chinese as Spacy nor GloVe handle it. The embeddings we will be using for Chinese are of dimension 100, therefore we need to adapt the embeddings for english from the dim 300 to 100. Glove does have English embeddings of dim 100 but Spacy does not have that model. So, we will tokenize the sentences using Spacy tokenizer and use GloVe directly.

#### Pre-processing English with GloVe

With GloVe's function *stoi()* (string to int) we can get the index corresponding to a given word and with the function *itos()* we get the word given its index. To obtain the vector of a word we first get the integer associated with it and then index it into the word embedding tensor with that index. Note that glove takes words in a lower case.

In [18]:
# DON'T RUN IF YOU ALREADY RAN IT IN THE ENGLISH-GERMAN SECTION
# Downloading spacy models for english

!spacy download en_core_web_md
!spacy link en_core_web_md en300

✔ Download and installation successful
You can now load the model via spacy.load('en_core_web_md')

✘ Link 'en300' already exists
To overwrite an existing link, use the --force flag



In [19]:
import torchtext
import spacy

#Embeddings
glove = torchtext.vocab.GloVe(name='6B', dim=100)

#tokenizer model
nlp_en =spacy.load('en300')



.vector_cache/glove.6B.zip: 862MB [06:28, 2.22MB/s]                          
100%|█████████▉| 399054/400000 [00:33<00:00, 22369.37it/s]

We can now write our functions that will return the average embeddings for a sentence.

In [20]:
#ENGLISH EMBEDDINGS methods from the section GERMAN-ENGLISH
# The difference from previous section is that we will use Glove embeddings directly because we are using a smaller model that spacy doesn't have
# We add a method to compute the word embedding and a method to compute the sentence embedding by averaging the word vectors

import numpy as np
import torch
from nltk import download
from nltk.corpus import stopwords

#downloading stopwords from the nltk package
download('stopwords') #stopwords dictionary, run once
stop_words_en = set(stopwords.words('english'))


def preprocess(sentence,nlp):
    text = sentence.lower()
    doc = [token.lemma_ for token in  nlp.tokenizer(text)]
    doc = [word for word in doc if word not in stop_words_en]
    doc = [word for word in doc if word.isalpha()] #restricts string to alphabetic characters only
    return doc

def get_word_vector(embeddings, word):
    try:
        vec = embeddings.vectors[embeddings.stoi[word]]
        return vec
    except KeyError:
        #print(f"Word {word} does not exist")
        pass

def get_sentence_vector(embeddings,line):
    vectors = []
    for w in line:
        emb = get_word_vector(embeddings,w)
        #do not add if the word is out of vocabulary
        if emb is not None:
            vectors.append(emb)
   
    return torch.mean(torch.stack(vectors))


def get_embeddings(f,embeddings,lang):
    file = open(f) 
    lines = file.readlines() 
    sentences_vectors =[]

    for l in lines:
        sentence= preprocess(l,lang)
        try:
            vec = get_sentence_vector(embeddings,sentence)
            sentences_vectors.append(vec)
        except:
            sentences_vectors.append(0)

    return sentences_vectors

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


#### Loading Chinese Word2Vec Embeddings

We now have to download the pre-trained embeddings for Chinese. We will get them from the University of Oslo NLPL repository (http://vectors.nlpl.eu/repository/), which has word2vec vectors of dimension 100.

 We will also get Chinese stop words from https://github.com/Tony607/Chinese_sentiment_analysis.

For embeddings of dimensions 300 you can find them searching on github repositories. One example is https://github.com/Kyubyong/wordvectors.

If you want to work on colab and download other embeddings I would suggest you download the file and upload it on your dropbox and get the link from there.



In [21]:
!wget -c https://github.com/Tony607/Chinese_sentiment_analysis/blob/master/data/chinese_stop_words.txt

!wget -O zh.zip http://vectors.nlpl.eu/repository/20/35.zip

!unzip zh.zip 

--2020-02-07 08:49:53--  https://github.com/Tony607/Chinese_sentiment_analysis/blob/master/data/chinese_stop_words.txt
Resolving github.com (github.com)... 192.30.253.113
Connecting to github.com (github.com)|192.30.253.113|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/html]
Saving to: ‘chinese_stop_words.txt’

chinese_stop_words.     [ <=>                ] 419.42K  --.-KB/s    in 0.06s   

2020-02-07 08:49:54 (6.55 MB/s) - ‘chinese_stop_words.txt’ saved [429483]

--2020-02-07 08:49:55--  http://vectors.nlpl.eu/repository/20/35.zip
Resolving vectors.nlpl.eu (vectors.nlpl.eu)... 129.240.189.225
Connecting to vectors.nlpl.eu (vectors.nlpl.eu)|129.240.189.225|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1458485917 (1.4G) [application/zip]
Saving to: ‘zh.zip’

zh.zip              100%[===================>]   1.36G  23.1MB/s    in 64s     

2020-02-07 08:51:00 (21.6 MB/s) - ‘zh.zip’ saved [1458485917/1458485917]

Arch

We now load the pre-trained word2vec embeddings we downloaded using the gensim package. More info on gensim and how to use it to load models and embeddings here https://radimrehurek.com/gensim/models/word2vec.html

In [22]:
from gensim.models import Word2Vec
from gensim.models import KeyedVectors

wv_from_bin = KeyedVectors.load_word2vec_format("model.bin", binary=True) 

/usr/local/lib/python3.6/dist-packages/smart_open/smart_open_lib.py:402: UserWarning: This function is deprecated, use smart_open.open instead. See the migration notes for details: https://github.com/RaRe-Technologies/smart_open/blob/master/README.rst#migrating-to-the-new-open-function
  'See the migration notes for details: %s' % _MIGRATION_NOTES_URL


#### Pre-processing Chinese

For pre-processing chinese sentence we will use the tokenizer package for chinese called jieba and use the downloaded list of chinese stop words to remove them from our tokens. More info on jieba and its options at https://github.com/fxsjy/jieba. 

In [ ]:
import string
import jieba
import gensim 
import spacy
import numpy as np

stop_words = [ line.rstrip() for line in open('./chinese_stop_words.txt',"r", encoding="utf-8") ]


def get_sentence_vector_zh(line):
    vectors = []
    for w in line:
        try:
            emb = wv_from_bin[w]
            vectors.append(emb)
        except:
            pass #Do not add if the word is out of vocabulary
    if vectors:
        vectors = np.array(vectors)
        return np.mean(vectors)  
    else:
        return 0


def processing_zh(sentence):
    seg_list = jieba.lcut(sentence,cut_all=True)
    doc = [word for word in seg_list if word not in stop_words]
    docs = [e for e in doc if e.isalnum()]
    return docs


def get_sentence_embeddings_zh(f):
    file = open(f) 
    lines = file.readlines() 
    sentences_vectors =[]
    for l in lines:
        sent  = processing_zh(l)
        vec = get_sentence_vector_zh(sent)

        if vec is not None:
            sentences_vectors.append(vec)
        else:
            print(l)
    return sentences_vectors

In [24]:
import spacy
import torchtext
from torchtext import data


zh_train_mt = get_sentence_embeddings_zh("./train.enzh.mt")
zh_train_src = get_embeddings("./train.enzh.src",glove,nlp_en)
f_train_scores = open("./train.enzh.scores",'r')
zh_train_scores = f_train_scores.readlines()


zh_val_src = get_embeddings("./dev.enzh.src",glove,nlp_en)
zh_val_mt = get_sentence_embeddings_zh("./dev.enzh.mt")
f_val_scores = open("./dev.enzh.scores",'r')
zh_val_scores = f_val_scores.readlines()

Building prefix dict from the default dictionary ...
Dumping model to file cache /tmp/jieba.cache
Loading model cost 0.997 seconds.
Prefix dict has been built successfully.


In [25]:
print(f"Training mt: {len(zh_train_mt)} Training src: {len(zh_train_src)}")
print()
print(f"Validation mt: {len(zh_val_mt)} Validation src: {len(zh_val_src)}")

Training mt: 7000 Training src: 7000

Validation mt: 1000 Validation src: 1000


In [ ]:
import numpy as np


X_train= [np.array(zh_train_src),np.array(zh_train_mt)]
X_train_zh = np.array(X_train).transpose()

X_val = [np.array(zh_val_src),np.array(zh_val_mt)]
X_val_zh = np.array(X_val).transpose()

#Scores
train_scores = np.array(zh_train_scores).astype(float)
y_train_zh =train_scores

val_scores = np.array(zh_val_scores).astype(float)
y_val_zh =val_scores


### Training the Regressor


At this point,  will try SVM and Random Tree Forests and choose the model with the highest Pearson correlation.

First we will define our RMSE function.

In [ ]:
import numpy as np

def rmse(predictions, targets):
    return np.sqrt(((predictions - targets) ** 2).mean())

#### SVM

SVM have many parameters such as the kernel and the regularizating constant C. Here we will use default C = 1 and compare kernels. 

In [28]:
from sklearn.svm import SVR
from scipy.stats.stats import pearsonr

for k in ['linear','poly','rbf','sigmoid']:
    clf_t = SVR(kernel=k)
    clf_t.fit(X_train_zh, y_train_zh)
    print(k)
    predictions = clf_t.predict(X_val_zh)
    pearson = pearsonr(y_val_zh, predictions)
    print(f'RMSE: {rmse(predictions,y_val_zh)} Pearson {pearson[0]}')
    print()


linear
RMSE: 0.9569251357826374 Pearson 0.078828752924285

poly
RMSE: 0.9573340665924616 Pearson 0.06684940857693276

rbf
RMSE: 0.9515605754342689 Pearson 0.11468776355641784

sigmoid
RMSE: 318.5385411750097 Pearson 0.061956801685699746



In this case, the radial basis function kernel performed the best with a Pearson correlation of 0.1147. 

#### Random Tree Forest

Another powerful regressor is the Random Tree Forest. Here we have to choose the number of trees we want to compute and we will pick n_estimators = 1000. The higher the number the longer it will compute. To fine tune that number you could compute the error per number of trees and select the number for which there is no more significant improvement( the "elbow" of the graph).

In [29]:
# Import the model we are using

from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(n_estimators=1000, random_state=666)
rf.fit(X_train_zh, y_train_zh);
predictions = rf.predict(X_val_zh)

pearson = pearsonr(y_val_zh, predictions)
print('RMSE:', rmse(predictions,y_val_zh))
print(f"Pearson {pearson[0]}")

RMSE: 0.9603546068992237
Pearson 0.0748575846434103


Finally, we see that SVM with RBF kernel is the best model here. We will now use it to predict on the test set.

### Writing Results

In [ ]:
import os

def writeScores(method_name,scores):
    fn = "predictions.txt"
    print("")
    with open(fn, 'w') as output_file:
        for idx,x in enumerate(scores):
            #out =  metrics[idx]+":"+str("{0:.2f}".format(x))+"\n"
            #print(out)
            output_file.write(f"{x}\n")
  

In [ ]:
#EN_ZH

zh_test_mt = get_sentence_embeddings_zh("./test.enzh.mt")
zh_test_src = get_embeddings("./test.enzh.src",glove,nlp_en)

X= [np.array(zh_test_mt),np.array(zh_test_src)]
X_test_zh = np.array(X).transpose()

#Predict
clf_zh = SVR(kernel='rbf')
clf_zh.fit(X_train_zh, y_train_zh)

predictions_zh = clf_zh.predict(X_test_zh)


In [34]:
#EN_ZH

from google.colab import files
from zipfile import ZipFile


writeScores("SVR",predictions_zh)

with ZipFile("en-zh_svr.zip","w") as newzip:
    newzip.write("predictions.txt")
 
files.download('en-zh_svr.zip')

### Results

Once submitted to codalab, the pearson correlation is 0.0795